client-sdk/ts-web/core/src/hdkey.ts

```diff
@@ -0,0 +1,93 @@
+import {sha512} from 'js-sha512';
+import {SignKeyPair, sign} from 'tweetnacl';
+import {generateMnemonic, mnemonicToSeed} from 'bip39';
+
+const ED25519_CURVE = 'ed25519 seed';
+const HARDENED_OFFSET = 0x80000000;
+const replaceDerive = (val: string): string => val.replace("'", '');
+const pathRegex = new RegExp("^m(\\/[0-9]+')+$");
+
+export class HDKey {
+    public readonly keypair: SignKeyPair;
+
+    constructor(public readonly privateKey: Uint8Array, public readonly chainCode: Uint8Array) {
+        this.keypair = sign.keyPair.fromSeed(privateKey);
+    }
+
+    /**
+     * Returns an instance of HDKey that can then be used for derivation
+     * @param seed Master seed for key derivation
+     * @returns Instance of HDKey
+     */
+    public static fromSeed(seed: Uint8Array): HDKey {
+        const hash = sha512.hmac.arrayBuffer(ED25519_CURVE, Buffer.from(seed));
+
+        const I = new Uint8Array(hash);
+        const IL = I.slice(0, 32);
+        const IR = I.slice(32);
+
+        return new HDKey(IL, IR);
+    }
+
+    /**
+     * Returns the root node for the given mnemonic
+     * @param mnemonic Bip39 Mnemonic
+     * @param passphrase Optional passphrase
+     * @returns Root node
+     */
+    public static async fromMnemonic(mnemonic: string, passphrase?: string): Promise<HDKey> {
+        const seed = await mnemonicToSeed(mnemonic, passphrase);
+        return HDKey.fromSeed(seed);
+    }
+
+    /**
+     * Generates a mnemonic
+     * @param strength Length in bits of the generated mnemonic
+     * @returns Generated BIP39 Mnemonic
+     */
+    public static generateMnemonic(strength = 256): string {
+        return generateMnemonic(strength);
+    }
+
+    /**
+     * Returns the HDKey for the given derivation path
+     * @param path Derivation path, starting with m/
+     * @returns Instance of HDKey
+     */
+    public derivePath(path: string): HDKey {
+        if (!path.match(pathRegex)) {
+            throw new Error(
+                "Invalid derivation path. Valid paths must use a format similar to : m/44'/474'/0' and all indexes must be hardened",
+            );
+        }
+
+        const segments = path
+            .split('/')
+            .slice(1)
+            .map(replaceDerive)
```

Comment:
I was going to suggest making this an explicit arrow fn just in case `replaceDerive` ever gets a second argument, but I saw that it's a simple function that should just be inlined.